experimental/fast_lossless/fast_lossless.cc

```diff
@@ -886,17 +957,156 @@ void CollectSamples(const unsigned char* rgba, size_t x0, size_t y0,
+  for (; i < pcolors; i++) {
+    if (i < pcolors_real) {
+      prgba = reinterpret_cast<const uint8_t*>(&palette[i]);
+    }
+    pdata[0][16 + i + have_zero] = prgba[0];
+    pdata[1][16 + i + have_zero] = prgba[1];
+    pdata[2][16 + i + have_zero] = prgba[2];
+    pdata[3][16 + i + have_zero] = prgba[3];
+  }
+  pdata[0][15] = 0;
+  row_encoder.ProcessRow(pdata[0] + 16, pdata[0] + 15, pdata[0] + 15,
+                         pdata[0] + 15, pcolors);
+  pdata[1][15] = pdata[0][16];
+  pdata[0][15] = pdata[0][16];
+  row_encoder.ProcessRow(pdata[1] + 16, pdata[1] + 15, pdata[0] + 16,
+                         pdata[0] + 15, pcolors);
+  pdata[2][15] = pdata[1][16];
+  pdata[1][15] = pdata[1][16];
+  row_encoder.ProcessRow(pdata[2] + 16, pdata[2] + 15, pdata[1] + 16,
+                         pdata[1] + 15, pcolors);
+  pdata[3][15] = pdata[2][16];
+  pdata[2][15] = pdata[2][16];
+  row_encoder.ProcessRow(pdata[3] + 16, pdata[3] + 15, pdata[2] + 16,
+                         pdata[2] + 15, pcolors);
+  row_encoder.Finalize();
+
+  if (!is_single_group) {
+    output->ZeroPadToByte();
+  }
+}
+
 size_t FastLosslessEncode(const unsigned char* rgba, size_t width,
                           size_t row_stride, size_t height,
                           unsigned char** output) {
   assert(width != 0);
   assert(height != 0);
   assert(row_stride >= 4 * width);
 
+  // Count colors to try palette
+  std::vector<uint32_t> palette(kMaxColors);
+  palette[0] = 1;
+  int16_t lookup[kMaxColors] = {};
+  int pcolors = 0;
+  bool collided = false;
+  for (size_t y = 0; y < height && !collided; y++) {
+    const unsigned char* r = rgba + row_stride * y;
+    size_t x = 0;
+    // this is just an unrolling of the next loop
+    for (; x + 3 < width; x += 4) {
+      uint32_t p[4], index[4];
+      memcpy(p, r + x * 4, 16);
+      for (int i = 0; i < 4; i++) {
+        index[i] = pixel_hash(p[i]);
+      }
+      for (int i = 0; i < 4; i++) {
+        uint32_t init_entry = index[i] ? 0 : 1;
+        if (init_entry != palette[index[i]] && p[i] != palette[index[i]]) {
+          collided = true;
+        }
+      }
+      for (int i = 0; i < 4; i++) palette[index[i]] = p[i];
+    }
+
+    for (; x < width; x++) {
+      uint32_t p;
+      memcpy(&p, r + x * 4, 4);
+      uint32_t index = pixel_hash(p);
+      uint32_t init_entry = index ? 0 : 1;
+      if (init_entry != palette[index] && p != palette[index]) {
+        collided = true;
+      }
+      palette[index] = p;
+    }
+  }
+
+  if (!collided) {
+    int k = 0;
+    if (palette[0] == 0) pcolors = 1;
+    if (palette[0] == 1) palette[0] = 0;
+    std::sort(palette.begin(), palette.end(), [](uint32_t a, uint32_t b) {
+      if (a == 0) return false;
+      if (b == 0) return true;
+      uint8_t a_rgba[4];
+      uint8_t b_rgba[4];
+      memcpy(a_rgba, &a, 4);
+      memcpy(b_rgba, &b, 4);
+      float a_y =
+          (0.299 * a_rgba[0] + 0.587 * a_rgba[1] + 0.114 * a_rgba[2] + 0.01) *
```

Comment:
probably want to do operations in floats (0.299f)

Reply:
indeed.